neutrinordp/xrdp-neutrinordp.c

```diff
@@ -609,6 +610,77 @@ lxrdp_set_param(struct mod *mod, const char *name, const char *value)
         {
             mod->perf_settings_values_mask |= PERF_DISABLE_CURSOR_SHADOW;
         }
+    }
+        else if (g_strcmp(name, "neutrinordp.allow_client_keyboardLayout") == 0)
+    {
+        if (g_text2bool(value))
+        {
+            if (settings->kbd_type == 0)
+            {
+                settings->kbd_type = mod->client_info.keyboard_type;
+            }
+            if (settings->kbd_subtype == 0)
+            {
+                settings->kbd_subtype = mod->client_info.keyboard_subtype;
+            }
+            if (settings->kbd_fn_keys == 0)
+            {
+                /* Define the most common number of function keys, 12. 
+                   because we can't get it from client. */
+                settings->kbd_fn_keys = 12;
+            }
+            if (settings->kbd_layout == 0)
+            {
+                settings->kbd_layout = mod->client_info.keylayout;
+                /* When the keyboard type is 0x07 (Japanese keyboard), 
+                   this function masks the upper 16 bits.*/
+                if (mod->client_info.keyboard_type == 0x07)
+                {
+                    LOG(LOG_LEVEL_DEBUG,
+                    "keyboardLayout mask for Japanese PC106/109 keyboard:"
+                    "[0x%08X]", 0x0000FFFF);
+                    settings->kbd_layout &= 0x0000FFFF;
+                }
+            }
+            LOG(LOG_LEVEL_INFO, "NeutrinoRDP proxy remote keyboard settings, "
+                                "kbd_type:[0x%02X], kbd_subtype:[0x%02X], "
+                                "kbd_fn_keys:[%02d], kbd_layout:[0x%08X]",
+                                settings->kbd_type, settings->kbd_subtype,
+                                settings->kbd_fn_keys, settings->kbd_layout);
+        }
+    }
+    /* The following options are for DEBUG and workaround.*/
+    else if (g_strcmp(name, "neutrinordp.override_keyboardLayout_masks") == 0)
```

Comment:
I think this code may behave differently depending on the order that these variables are encountered in the .ini file. If for example I put this:-

```
neutrinordp.allow_client_keyboardLayout=true
neutrinordp.override_kbd_type=0
```

I'll get a `settings->kbd_type` of zero.

However if I put this:-
```
neutrinordp.override_kbd_type=0
neutrinordp.allow_client_keyboardLayout=true
```

I'll get the client keyboard type.

Reply:
Thank you.

I forgot the case of zero, so the logic is broken....

We wanted to deal with this in a simple way, so as not to add more member variables to the mod structure than necessary.
For this reason, we thought of handling it within the `lxrdp_set_param()` function.
On the other hand, the behavior may change depending on the order in which the parameters are written.

Giving up on processing within the `lxrdp_set_param()` function, the `lfreerdp_pre_connect()` function?

Reply:
I have a problem here.

I've tested it on a local build, but
Depending on the keyboard layout that xrdp_wm is aware of, it may not work as intended. Sometimes it works.
I think it is only for debugging use, where someone who has read the source code and understands how it works can change the parameters and see the change in the remote Windows behavior.

remove it from the `xrdp.ini.in` example and
Leave it as an undocumented option for debug that can be used in any order of description.

Or should I just delete the code for the `neutrinordp.overrode_*` option?

Reply:
I think getting it working is the best solution, and I also think it can be got working with a bit of effort.

I think the problem is that you're trying to do all the work while the parameters are being processed in `lxrdp_set_param()`. The overrides need to be moved to after the loop where the parameters are processed. You can do this by setting the keyboard parameters before the connection is made.

Here's a rough example. Please tell me if I'm not being clear. Also I haven't tested this! I've made some simplifications to your code which may not work well, but hopefully will make the idea clearer.

- In xrdp-neutrinordp.h
```c
/* Values in the ini file used to override keyboard settings */
struct kbd_overrides
{
    int type;
    int subtype;
    int fn_keys;
    int layout;
    int layout_mask;
};
. . .
struct mod
{
. . .
    int allow_client_kbd_settings;
    struct kbd_overrides kbd_overrides; /* neutrinordp.override_kbd_* values */
};
```

- Then in xrdp-neutrinordp.c:-
    - in `lxrdp_set_params()`, just record values. Don't change anything.
    - add a new function called `set_keyboard_overrides()`
    - call `set_keyboard_overrides()` from `lxrdp_connect()` after all the parameters have been read.

So the code in `lxrdp_set_params()` will look a bit like this:-
```c
    else if (g_strcmp(name, "neutrinordp.allow_client_keyboardLayout") == 0)
    {
        /* Keyboard values are stored for later processing */
        mod->allow_client_kbd_settings = g_text2bool(value);
    }
    else if (g_strcmp(name, "neutrinordp.override_keyboardLayout_mask") == 0)
    {
        mod->kbd_overrides.layout_mask = g_atoi(value);
    }
    else if (g_strcmp(name, "neutrinordp.override_kbd_type") == 0)
    {
        mod->kbd_overrides.type = g_atoi(value);
    }
    else if (g_strcmp(name, "neutrinordp.override_kbd_subtype") == 0)
    {
        mod->kbd_overrides.subtype = g_atoi(value);
    }
    else if (g_strcmp(name, "neutrinordp.override_kbd_fn_keys") == 0)
    {
        mod->kbd_overrides.fn_keys = g_atoi(value);
    }
    else if (g_strcmp(name, "neutrinordp.override_kbd_layout") == 0)
    {
        mod->kbd_overrides.layout = g_atoi(value);
    }
```

And the function `set_keyboard_overrides()` will look a bit like this.

```c
set_keyboard_overrides(struct mod *mod)
{
    const struct kbd_overrides *ko = &mod->kbd_overrides;
    rdpSettings *settings  = settings = mod->inst->settings;

    if (mod->allow_client_kbd_settings)
    {
        settings->kbd_type = mod->client_info.keyboard_type;
        settings->kbd_subtype = mod->client_info.keyboard_subtype;
        /* Define the most common number of function keys, 12. 
           because we can't get it from client. */
        settings->kbd_fn_keys = 12;

        settings->kbd_layout = mod->client_info.keylayout;
        /* When the keyboard type is 0x07 (Japanese keyboard), 
           this function masks the upper 16 bits.*/
        if (mod->client_info.keyboard_type == 0x07)
        {
            LOG(LOG_LEVEL_DEBUG,
            "keyboardLayout mask for Japanese PC106/109 keyboard:"
            "[0x%08X]", 0x0000FFFF);
            settings->kbd_layout &= 0x0000FFFF;
        }

        LOG(LOG_LEVEL_INFO, "NeutrinoRDP proxy remote keyboard settings, "
                            "kbd_type:[0x%02X], kbd_subtype:[0x%02X], "
                            "kbd_fn_keys:[%02d], kbd_layout:[0x%08X]",
                            settings->kbd_type, settings->kbd_subtype,
                            settings->kbd_fn_keys, settings->kbd_layout);
    }


    if (ko->type != 0)
    {
        LOG(LOG_LEVEL_INFO, "overrode kbd type 0x%02X with 0x%02X",
            settings->kbd_type, ko->type);
        settings->kbd_type = ko->type;
    }

    if (ko->subtype != 0)
    {
        LOG(LOG_LEVEL_INFO, "overrode kbd subtype 0x%X with 0x%X",
            settings->kbd_subtype, ko->subtype);
        settings->kbd_subtype = ko->subtype;
    }

    if (ko->fn_keys != 0)
    {
        LOG(LOG_LEVEL_INFO, "overrode kbd fn keys 0x%X with 0x%X",
            settings->kbd_fn_keys, ko->fn_keys);
        settings->kbd_fn_keys = ko->fn_keys;
    }

    if (ko->layout != 0)
    {
        LOG(LOG_LEVEL_INFO, "overrode kbd layout 0x%X with 0x%X",
            settings->kbd_layout, ko->layout);
        settings->kbd_layout = ko->layout;
    }

    if (ko->layout_mask != 0)
    {
        LOG(LOG_LEVEL_INFO, "Masked kbd layout 0x%X to 0x%X",
            settings->kbd_layout, settings->kbd_layout & ko->layout_mask);
        settings->kbd_layout &= ko->layout_mask;
    }
}
```

Hope that's useful.

Reply:
@matt335672 Thank you.
We will code and test it based on your suggestions.

Due to my work, please wait until next week for the result.